### PR TITLE
Analytical IBS growth rates tailored to Xsuite

### PR DESCRIPTION
**Pull request by fsoubelet**
## Description

To be synced with https://github.com/xsuite/xtrack/pull/478

**New:**

This PR adds an `ibs` sub-package to `xfields`, in which I have implemented the analytical calculations for IBS growth rates, ported with heavy modifications from my work in the `xibs` prototype. All modules are private, and the desired API we discussed is defined in `_api.py` and exported to the user at the module level in `__init__.py`. 

Some helper functions to compute properties from `xtrack.Particles` are in a `_formulary.py` file, and all "die gracefully" if the user is on the PyOpenCL context since it gives wrong results with boolean array masking.

Core functionality is in the `_analytical.py` file. An ABC class holds some common functionality and defines the API any implementation needs to follow, should we want to add some more later and have them work out-of-the-box with the exported API. Classes instantiate from a `TwissTable`. The exposed API accepts geometric and normalized emittances, but also simply passing an `xtrack.Particles` object. Implemented formalism are Nagaitsev (https://journals.aps.org/prab/pdf/10.1103/PhysRevSTAB.8.064403) and Bjorken-Mtingwa (the revision that accounts for non-relativistic beams and vertical dispersion, see https://cds.cern.ch/record/1445924/files/CERN-ATS-2012-066.pdf).

**Added examples:**
- One example to get the rates from given parameters, without Dy (needs https://github.com/xsuite/xtrack/pull/478)
- One example to get the rates from given parameters, _with_ Dy (needs https://github.com/xsuite/xtrack/pull/478)
- One example to get the rates from given particles (needs https://github.com/xsuite/xtrack/pull/478)
- All 3 of the above compute rates for Nagaitsev & Bjorken-Mtingawa and print them out

**Tests:**
- I have added an `ibs_conftest.py` file with a few helpers.
- I have added just a few tests as discussed in order to not introduce too much. They get growth rates from MAD-X and compare the ones I get. I have a test case for positive charge, negative charge, ions, and one with vertical dispersion.
- I parametrized to test for bunched and coasting beams.
- I do not test Nagaitsev vertical if Dy is present since the formalism does not account for it and gives wrong results.
- I do not test Nagaitsev rates if considering coasting beams as it makes some big assumptions. For a correct treatment the Bjorken-Mtingwa formalism is recommended (this is logged to the user).
- Currently on my laptop this adds 2min02s to test runs, or 39s if parallelized.

**Some miscellaneous points:**
- I have not yet looked into completely adapting the references in my docstrings (copied from xibs), would be happy with guidance.
- I have not adapted all admonitions in my docstrings and would be happy with guidance. 
- Structures are implemented as `xobjects.HybridClass`, which is useless for now but will become relevant once we want to speed-up part of the kicks code.
- I have bumped the version number to `0.16.0`
- I have shamelessly added myself to the contributors :D
- I am preparing an IBS section in the physics guide but I will work on the kicks PR first.

## Checklist

Mandatory: 

- [x] I have added tests to cover my changes
- [x] All the tests are passing, including my new ones
- [x] I described my changes in this PR description

Optional:

- [x] The code I wrote follows good style practices (see [PEP 8](https://peps.python.org/pep-0008/) and [PEP 20](https://peps.python.org/pep-0020/)).
- [ ] I have updated the docs in relation to my changes, if applicable
- [ ] I have tested also GPU contexts

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
+from pathlib import Path
```
Comment (giadarol) [inline tests/conftest.py:1]: Let's rename the file to `ibs_conftest.py`

File: tests/test_ibs_growth_rates.py [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+        gemitt_y=gemitt_y,
+        sigma_delta=sigd,
+        bunch_length=bl,
+        bunched=bunched,
+    )
+    # -----------------------------------------------------
+    # Compare the results - Nagaitsev
+    if bunched is True:  # in Nagaitsev coasting makes big assumptions
+        assert_allclose(nag_rates.Tx, mad_Tx, atol=1e-8, rtol=2.5e-2)
+        assert_allclose(nag_rates.Ty, mad_Ty, atol=1e-8, rtol=10e-2)
```
Comment (giadarol) [inline tests/test_ibs_growth_rates.py:137]: Let's add a comment saying that the 10% is not surprising since the model is different

**Review by giadarol — CHANGES_REQUESTED** [review]

File: tests/test_ibs_growth_rates.py (hunk 2 above)
Comment (fsoubelet) [reply tests/test_ibs_growth_rates.py:138]: Done

File: tests/conftest.py (hunk 1 above)
Comment (fsoubelet) [reply tests/conftest.py]: Done

**Comment by fsoubelet** [comment]
Changed the example files to get the growth rates directly using the new method from the `TwissTable`. This requires https://github.com/xsuite/xtrack/pull/478 to be merged so it can run.

**Review by giadarol — APPROVED** [review]

**Pull request merged**